src/coreclr/jit/layout.cpp

```diff
@@ -730,6 +742,80 @@ ClassLayoutBuilder::ClassLayoutBuilder(Compiler* compiler, unsigned size)
 {
 }
 
+//------------------------------------------------------------------------
+// BuildArray: Construct a builder for an array layout
+//
+// Arguments:
+//    compiler      - Compiler instance
+//    arrayHandle   - class handle for array
+//    length        - array length (in elements)
+//
+// Note:
+//    For arrays of structs we currently do not copy any struct padding,
+//    with the presumption that it is unlikely we will ever promote array elements.
+//
+ClassLayoutBuilder ClassLayoutBuilder::BuildArray(Compiler* compiler, CORINFO_CLASS_HANDLE arrayHandle, unsigned length)
+{
+    assert(length <= CORINFO_Array_MaxLength);
+    assert(arrayHandle != NO_CLASS_HANDLE);
+
+    CORINFO_CLASS_HANDLE elemClsHnd = NO_CLASS_HANDLE;
+    CorInfoType          corType    = compiler->info.compCompHnd->getChildType(arrayHandle, &elemClsHnd);
+    var_types            type       = JITtype2varType(corType);
+
+    ClassLayout* elementLayout = nullptr;
+    unsigned     elementSize   = 0;
+
+    if (type == TYP_STRUCT)
+    {
+        elementLayout = compiler->typGetObjLayout(elemClsHnd);
+        elementSize   = elementLayout->GetSize();
+    }
+    else
+    {
+        elementSize = genTypeSize(type);
```

Comment:
I suppose if we really wanted to we could provide a constructor for a class layout for primitive types, which could be used to unify paths like these.

Reply:
In any case we can just create them as custom layouts now, but I suppose it would be slightly less efficient than this approach.